### DISCUSSION TITLE
Split router definition and implementation

### DISCUSSION THREAD

**Discussion by unshame** [trpc/trpc#3657] [Q&A] [answered]
Is it possible to define the router type in one place and then provide query/mutation implementations for it in another? I want to lift the api definitions into a separate package, so that both client and server can import from it, instead of client importing from server directly. Is this doable?

**Comment by Nick-Lucas** [comment, marked as answer]
tRPC is designed to work the other way around, you can get interfaces from the implementation, but not an implementation from interfaces. Some people choose to define their routers in a library and then have a skinny API app which imports them from the library. 

Why are you worried about doing it this way? People who want to do it backwards are usually worried that importing the types from the API directly will inadvertently bundle the API into their client, but if you did accidentally do this then the client wouldn't even boot because it's using node APIs in the browser. By doing `import type { AppRouter } from 'path/to/api` you're guaranteed not to import any code from the backend, just the types for compile-time checks

**Reply by unshame** [reply]
Alright, that makes sense. The only overhead in doing it this way is that I will need to compile the server in order to compile the client, as opposed to simply compiling the api package. But that's probably fine.

**Reply by Nick-Lucas** [reply]
Makes sense as a concern, and there's definitely overhead to compiling more typescript. The performance of tRPC type checking since v10 is pretty good though and when I've had issues it's almost always been down to Zod types getting complex, which is easy enough to avoid by keeping them simple (no subtypes from extend, merge, pick, omit, etc)

**Reply by Akxe** [reply]
I for example do not like the fact, that all of the methods, validation and similar things have to be in the common library and if there would be any code leak, of faulty tree shaking, the implementation detail would leak to the userland. It looks like a security bomb
